### PR TITLE
jlm-hls: Tool for interacting with the libljm HLS backend

### PR DESCRIPTION
**Pull request by sjalander**
The tool has support extracting specified function for synthesis.
Inlining is performed such that the extracted function does not
contain any function calls. Thus, the LLVM IR used as input has
to contain all functions that's called by the extracted function.

The tool also has support for generating FIRRTL or a DOT graph of
the provided input function after the RVSDG has been created and
lowered to RHLS.

### CODE REVIEW

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <jlm/util/Statistics.hpp>
+
+#include <string>
+#include <vector>
+
+namespace jlm {
+
+class optimization;
+
+enum class outputformat {fir, dot};
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:19]: Could we rename fir to firrtl? It is more descriptive.

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+#ifndef JLM_JLMHLS_CMDLINE_HPP
+#define JLM_JLMHLS_CMDLINE_HPP
+
+#include <jlm/util/file.hpp>
+#include <jlm/util/Statistics.hpp>
+
+#include <string>
+#include <vector>
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:13]: Is the "#include <vector>" really needed? I do not see a std::vector used in this file?

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace jlm {
+
+class optimization;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:17]: You can remove this forward definition. It does not seem to be used. Copy-paste error?

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+class cmdline_options {
+public:
+	cmdline_options()
+	: ifile("")
+	, outfolder("")
+	, format(outputformat::fir)
+	{}
+
+	jlm::filepath ifile;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:29]: Could we rename it to inputFile? More descriptive.

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	jlm::filepath outfolder;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:30]: Rename to outputFolder?

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	outputformat format;
+	std::string hls_function;
+	bool extract = false;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:33]: Rename to extractKernel or extractHlsFunction?

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	bool circt = false;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:34]: Rename to useCirct?

File: jlm-hls/include/jlm-hls/cmdline.hpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	StatisticsDescriptor sd;
```
Comment (phate) [inline jlm-hls/include/jlm-hls/cmdline.hpp:35]: The StatisticsDescriptor is not set in the parse_cmdline function. No need to have it as an attribute here.

File: jlm-hls/src/cmdline.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	cl::opt<bool> circt(
+	  "circt"
+	, cl::Prefix
+	, cl::desc("Use CIRCT to generate FIRRTL"));
+
+	cl::opt<outputformat> format(
+	  cl::values(
+		  clEnumValN(outputformat::fir, "fir", "Output FIRRTL [default]")
+                , clEnumValN(outputformat::dot, "dot", "Output DOT graph"))
+        , cl::desc("Select output format"));
```
Comment (phate) [inline jlm-hls/src/cmdline.cpp:55]: Indentation!

File: jlm-hls/src/cmdline.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+	cl::ParseCommandLineOptions(argc, argv);
+
+	if (outfolder.empty()) {
+		std::cerr << "jlm-hls no output directory provided, i.e, -o.\n";
+                exit(EXIT_FAILURE);
```
Comment (phate) [inline jlm-hls/src/cmdline.cpp:61]: Identation!

File: jlm-hls/src/cmdline.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if (extract && hls_function.empty()) {
+		std::cerr << "jlm-hls: --hls-function is not specifided,\n";
+		std::cerr << "         which is required for --extract\n";
+                exit(EXIT_FAILURE);
```
Comment (phate) [inline jlm-hls/src/cmdline.cpp:67]: Identation!

File: jlm-hls/src/jlm-hls.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+#include <llvm/IR/Module.h>
+#include <llvm/IRReader/IRReader.h>
+#include <llvm/Support/raw_os_ostream.h>
+#include <llvm/Support/SourceMgr.h>
+
+static void
+dump_llvm(
+	  jlm::RvsdgModule &module,
+	  std::string outfile)
+{
```
Comment (phate) [inline jlm-hls/src/jlm-hls.cpp:31]: Identation!

File: jlm-hls/src/jlm-hls.cpp [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		std::string fileName = flags.outfolder.path() + "/jlm_hls.dot";
+		std::ofstream outFile;
+		outFile.open(fileName);
+		outFile << output;
+		outFile.close();
+	} else {
+		std::cerr << "This combination of option should never be reached.\n";
+		exit(EXIT_FAILURE);
+	}
+	return 0;
```
Comment (phate) [inline jlm-hls/src/jlm-hls.cpp:107]: if (flags.extract) {
  /* my extract code */
  return 0;
}

jlm::hls::rvsdg2hls(*rsvdgModule);

if (flags.format == jlm::outputformat::fir) {
  /* firrtl code */
} else if (flags.format == jlm::outputformat::dot) {
  /* dot code */
} else {
  JLM_UNREACHABLE("This combination of options should have never been reached.\n");
}

File: jlm-hls/src/jlm-hls.cpp [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+		} else {
+			jlm::hls::FirrtlHLS hls;
+			output = hls.run(*rvsdgModule);
+		}
+
+		std::string fileName = flags.outfolder.path() + "/jlm_hls.fir";
+		std::ofstream outFile;
+		outFile.open(fileName);
+		outFile << output;
+		outFile.close();
```
Comment (phate) [inline jlm-hls/src/jlm-hls.cpp:85]: Could we put this writing of an string to an output file wrap into its own function?

File: jlm-hls/src/jlm-hls.cpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+		jlm::hls::VerilatorHarnessHLS vhls;
+		output = vhls.run(*rvsdgModule);
+		fileName = flags.outfolder.path() + "/jlm_hls_harness.cpp";
+		outFile.open(fileName);
+		outFile << output;
+		outFile.close();
```
Comment (phate) [inline jlm-hls/src/jlm-hls.cpp:92]: Same here, please wrap into its own function.

File: jlm-hls/src/jlm-hls.cpp [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	} else if (flags.format == jlm::outputformat::dot) {
+		jlm::hls::rvsdg2rhls(*rvsdgModule);
+
+		jlm::hls::DotHLS dhls;
+		std::string output = dhls.run(*rvsdgModule);
+		std::string fileName = flags.outfolder.path() + "/jlm_hls.dot";
+		std::ofstream outFile;
+		outFile.open(fileName);
+		outFile << output;
+		outFile.close();
```
Comment (phate) [inline jlm-hls/src/jlm-hls.cpp:102]: And again.

File: libjlm/include/jlm/backend/hls/rhls2firrtl/base-hls.hpp [hunk 17]
```diff
@@ -18,19 +18,13 @@ namespace jlm {
...
 				create_node_names(get_hls_lambda(rm)->subregion());
 				auto text = get_text(rm);
-				std::basic_string<char, std::char_traits<char>, std::allocator<char>> base_file_name = get_base_file_name(
-						rm);
-				std::string file_name = base_file_name + extension();
-				std::ofstream out_file;
-				out_file.open(file_name);
-				out_file << text;
-				out_file.close();
+				return text;
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rhls2firrtl/base-hls.hpp:27]: return get_text(rm);

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp [hunk 18]
```diff
@@ -45,7 +45,7 @@ namespace jlm {
 			MLIRGenImpl(mlir::MLIRContext &context) : builder(&context) {}
 			circt::firrtl::CircuitOp MlirGen(const jlm::lambda::node *lamdaNode);
 			void WriteModuleToFile(const circt::firrtl::FModuleOp fModuleOp, const jive::node *node);
-			void WriteCircuitToFile(const circt::firrtl::CircuitOp circuit, std::string name);
+			std::string WriteCircuitToFile(const circt::firrtl::CircuitOp circuit, std::string name);
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp:48]: Could we maybe rename this function to something else as it does not write anything to a file any longer?

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp [hunk 19]
```diff
@@ -227,9 +227,11 @@ namespace jlm {
                         }
                         std::string get_text(jlm::RvsdgModule &rm) override {return "";}
                 public:
-                        void
+			std::string
                         run(jlm::RvsdgModule &rvsdgModule) {
                                 std::cout << "This version of jlm-hls has not been compiled with support for the CIRCT backend\n";
+				exit(EXIT_FAILURE);
+				return "";
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp:234]: Identation!

Also, please throw an exception:

throw jlm::error("This version of jlm-hls has not been compiled with CIRCT support.");

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 1 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: Removed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 2 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: string and vector have been removed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 3 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: forward definition has been removed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 4 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: renamed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 5 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: renamed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 6 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: renamed to extractHlsFunction

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 7 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: renamed

File: jlm-hls/src/jlm-hls.cpp (hunk 16 above)
Comment (sjalander) [reply jlm-hls/src/jlm-hls.cpp]: replace with stringToFile()

File: libjlm/include/jlm/backend/hls/rhls2firrtl/base-hls.hpp (hunk 17 above)
Comment (sjalander) [reply libjlm/include/jlm/backend/hls/rhls2firrtl/base-hls.hpp]: Changed

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp (hunk 19 above)
Comment (sjalander) [reply libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp]: Changed

File: libjlm/src/backend/hls/rhls2firrtl/mlirgen.cpp [hunk 20]
```diff
@@ -2372,17 +2372,18 @@ jlm::hls::MLIRGenImpl::WriteCircuitToFile(const circt::firrtl::CircuitOp circuit
...
 
 	std::string fileName = name + extension();
 	std::error_code EC;
 
-	llvm::raw_fd_ostream output(fileName, EC);
+	std::string outputString;
+	llvm::raw_string_ostream output(outputString);
 	auto status = circt::firrtl::exportFIRFile(module, output);
 	if (status.failed())
 		throw std::logic_error("Exporting of firrtl failed");
```
[the remark this answers is not in the document]
Comment (sjalander) [reply libjlm/src/backend/hls/rhls2firrtl/mlirgen.cpp:2405]: Changed

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp (hunk 18 above)
Comment (sjalander) [reply libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen.hpp]: This function is used in other places, so I've created a new called toString() instead.

File: jlm-hls/src/jlm-hls.cpp (hunk 13 above)
Comment (sjalander) [reply jlm-hls/src/jlm-hls.cpp]: Changed

File: jlm-hls/src/jlm-hls.cpp (hunk 15 above)
Comment (sjalander) [reply jlm-hls/src/jlm-hls.cpp]: Replaced with stringToFile()

File: jlm-hls/src/jlm-hls.cpp (hunk 14 above)
Comment (sjalander) [reply jlm-hls/src/jlm-hls.cpp]: Replaced with stringToFile()

File: jlm-hls/src/jlm-hls.cpp (hunk 12 above)
Comment (sjalander) [reply jlm-hls/src/jlm-hls.cpp]: Renamed to llvmToFile() and changed indentation

File: jlm-hls/src/cmdline.cpp (hunk 11 above)
Comment (sjalander) [reply jlm-hls/src/cmdline.cpp]: Changed

File: jlm-hls/src/cmdline.cpp (hunk 10 above)
Comment (sjalander) [reply jlm-hls/src/cmdline.cpp]: Changed

File: jlm-hls/src/cmdline.cpp (hunk 9 above)
Comment (sjalander) [reply jlm-hls/src/cmdline.cpp]: Changed

File: jlm-hls/include/jlm-hls/cmdline.hpp (hunk 8 above)
Comment (sjalander) [reply jlm-hls/include/jlm-hls/cmdline.hpp]: Removed

**Pull request merged**